src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/CounterfactualLogger.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading;
+using NuGet.Common;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Keeps all counterfactuals state in one place
+    /// </summary>
+    internal class CounterfactualLogger
+    {
+        private int _emittedFlag = 0;
+
+        internal CounterfactualLogger(string eventName)
+        {
+            EventName = eventName + "Counterfactual";
+        }
+
+        internal string EventName { get; }
+
+        /// <summary>
+        /// Emits counterfactual telemetry event once
+        /// </summary>
+        public void TryEmit()
```

Comment:
As per my understanding, `Try*` methods don't throw an exception but returns a `boolean` to notify the status of the operation to the caller. For e.g. https://docs.microsoft.com/en-us/dotnet/api/system.threading.monitor.tryenter?view=net-6.0, `If the lock was not taken because an exception was thrown, the variable specified for the lockTaken parameter is false after this method ends`. What does everyone think?

Reply:
If I understood correctly Interlocked.CompareExchange is bit different from traditional lock, it doesn't use above `Monitor` internally.

Reply:
Ok, I will rename it to avoid confusion.